### PR TITLE
fix(mlx): support text label parity datasets

### PR DESCRIPTION
**Pull request by Lyxot**
## Summary

Adds MLX text dataset label parity for CUDA/TRL-style labeled text training:

- text `prompt` / `completion` rows with `completion_only_loss`
- ordered, streaming, train, and eval text batch paths
- pre-tokenized text rows with `input_ids`, optional `labels`, `completion_mask`, and `assistant_masks`
- conversational text rows with `assistant_only_loss` / chat-template `assistant_masks`

The branch is scoped to text label construction and batching. It does not change packing, custom collators, VLM filtering, save/export, or broader sampler policy.

## What Changed

- Builds labeled MLX text batches for prompt/completion datasets instead of training on the full prompt+completion sequence.
- Mirrors TRL's default `completion_only_loss=None` behavior: prompt/completion rows train on completion labels, plain LM rows train on the full sequence.
- Supports conversational prompt/completion boundaries with `apply_chat_template(..., add_generation_prompt=True)`.
- Threads text `completion_only_loss` through default, ordered, streaming, and eval batch creation.
- Supports pre-tokenized rows without re-tokenizing `input_ids`.
- Uses explicit `labels` as base labels when present.
- Applies `completion_mask` only when completion-only loss resolves true.
- Applies `assistant_masks` as label metadata for pre-tokenized rows, matching TRL's collator behavior.
- Adds text `assistant_only_loss` support for chat-template assistant masks.
- Rejects assistant-only inputs that cannot produce assistant labels instead of silently training full text or dropping rows.
- Keeps labeled batches using the existing MLX length-sorted/shuffled or explicit ordered/streaming paths.

## Root Cause

MLX text batching previously delegated most text datasets through the generic mlx-lm text path, which emits `labels=None`. That lost CUDA/TRL label semantics for prompt/completion rows, already-tokenized rows, completion masks, and assistant masks, so MLX could train with different effective labels than CUDA.

## CUDA/TRL Parity Notes

This mirrors TRL `SFTTrainer` / `DataCollatorForLanguageModeling` behavior for supported text rows:

- `completion_only_loss=None` resolves true for prompt/completion rows and false for language-modeling rows.
- `formatting_func` is incompatible when completion-only loss is active.
- `labels` are used as the base labels when provided.
- `completion_mask` and `assistant_masks` mask labels to `-100` after base labels are selected.
- Padding labels use `-100`.
- Prompt/completion completion text receives EOS when needed.
- Rows with no supervised completion tokens after truncation are skipped before batching.
- Pre-tokenized `assistant_masks` are honored as collator metadata.

MLX also fails loudly for assistant-only rows that cannot produce assistant labels, avoiding silent full-label training or row drops.

## Diff / Review Size

Current branch diff against `main`:

```text
3 files changed, 1288 insertions(+), 24 deletions(-)
tests: 1 file changed, 450 insertions(+)
```

Test code ratio by added lines: `450 / 1288 = 34.94%`.

Commit stack:

```text
c88503af fix(mlx): support text completion-only batches
9e07a169 fix(mlx): support ordered completion-only batches
42e4f445 fix(mlx): support streaming completion-only batches
7c158cc4 fix(mlx): support pretokenized text batches
de7f4e4c fix(mlx): keep empty pretokenized rows processed
3668056f fix(mlx): support text assistant-only labels
```

## Validation

- `python -m py_compile tests/test_mlx_trainer_internals.py unsloth_zoo/mlx/utils.py unsloth_zoo/mlx/trainer.py`
- `python -m pytest tests/test_mlx_trainer_internals.py -q -k 'completion_only or pretokenized or assistant or prepare_data'`
  - `17 passed, 53 deselected`
- `git diff --check`

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/utils.py [hunk 1]
```diff
@@ -3548,10 +3549,351 @@ def __len__(self):
...
+                tokenizer, item
+            )
+        else:
+            prompt_text, completion_text = _prompt_completion_text_parts(
+                tokenizer, item, dataset_text_field
+            )
+            if eos_token and not completion_text.endswith(eos_token):
+                completion_text += eos_token
+            input_ids = _to_int_list(encode_mlx_text(tokenizer, prompt_text + completion_text))
+            prompt_ids = _to_int_list(encode_mlx_text(tokenizer, prompt_text))
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/utils.py]: Verified against TRL/SFTTrainer: the CUDA path appends EOS to prompt/completion text by checking the string suffix before tokenization. I kept the MLX path string-based here for CUDA parity rather than switching this PR to token-id EOS handling.

File: unsloth_zoo/mlx/utils.py [hunk 2]
```diff
@@ -3548,10 +3549,351 @@ def __len__(self):
...
+        input_ids = input_ids[:max_seq_length]
+        if len(input_ids) < 2:
+            continue
+
+        row = {"input_ids": input_ids}
+        if mask_completions:
+            prompt_len = min(len(prompt_ids), len(input_ids))
+            row["completion_mask"] = [0] * prompt_len + [1] * (len(input_ids) - prompt_len)
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/utils.py]: Checked against the current TRL/SFTTrainer implementation: CUDA warns on a prompt/prompt+completion tokenizer-prefix mismatch but still builds the completion mask from len(prompt_ids); it does not back off the prompt length. Keeping the MLX behavior aligned with CUDA here.

File: unsloth_zoo/mlx/utils.py [hunk 3]
```diff
@@ -3548,10 +3549,351 @@ def __len__(self):
...
+        raise ValueError(
+            "Unsloth MLX: prompt/completion dataset produced no trainable token "
+            "sequences (need at least two input_ids after truncation)."
+        )
+    return rows
+
+
+def _make_labeled_text_batch_tuple(batch_rows, max_seq_length, pad_id=0):
+    """Collate normalized text rows into an MLX batch with effective labels."""
+    max_len = min(max(len(row["input_ids"]) for row in batch_rows), max_seq_length)
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/utils.py]: The public create_batches path already slices mlx-lm 32-padded plain-text batches back to the true batch max length before returning. I kept the new labeled path consistent with that public return shape, fixed the default length-sorted/shuffled grouping separately, and updated the docstring to make this explicit.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Reviewed and pushed fixes for the genuine correctness issues Codex surfaced over the last few rounds:

- Drop labeled rows whose completion is entirely `-100` after truncation (a fully-masked batch would abort training mid-run), and honor `seed=0` as a deterministic seed.
- Raise for explicit `completion_only_loss=True` when a materialized or ordered text dataset yields no prompt/completion rows, matching the streaming path instead of silently training on the full sequence.
- Accept pretokenized `assistant_only_loss` rows that already carry explicit `labels` (without `assistant_masks`), and validate provided `assistant_masks` so all-zero masks fail loudly instead of being silently dropped.
- Drop pretokenized rows under two tokens (the labels-free baseline loss divided by zero), and accept single `{role, content}` prompt/completion messages in the assistant-only prevalidation so valid rows the tokenizer path already supports are not rejected.

Two remaining Codex items I left for you as design calls rather than change directly:

- `formatting_func` supplied together with prompt/completion data under `completion_only_loss=None` (utils.py:2855): the row is rendered before prompt/completion detection runs, so it trains full-text. Your call whether to reject that combination or resolve labels from the raw row before applying the formatter.
- `completion_only_loss=True` combined with `assistant_only_loss=True` on a conversational dataset (utils.py:2869): the assistant-only fallback currently wins and bypasses the completion-only guard. Up to you whether to raise on the conflicting pair or define an explicit precedence.

Everything else looks good.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/utils.py [hunk 4]
```diff
@@ -3569,6 +4225,62 @@ def create_batches(dataset, tokenizer, batch_size, max_seq_length,
...
+        return _create_tokenized_text_batches(
+            tokenized,
+            batch_size,
+            max_seq_length,
+            num_batches=num_batches,
+            seed=seed,
+            pad_id=_mlx_text_pad_id(tokenizer),
+        )
+    if assistant_only_loss:
+        _validate_mlx_text_assistant_only_dataset(dataset)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py:4656]: This mirrors TRL exactly and is intended parity. TRL runs the same assistant_only_loss conversational precheck on the raw sample before applying formatting_func (is_conversational is resolved on the pre-format dataset_sample and raises the same error). TRL's formatting_func returns text, not conversational messages, and its post-format language-modeling branch produces no assistant masks either, so rejecting the raw non-conversational dataset here is correct parity rather than a bug. Leaving as-is.

File: unsloth_zoo/mlx/trainer.py [hunk 5]
```diff
@@ -2047,6 +2056,7 @@ def _prepare_data(self, is_vlm):
             if args.max_steps > 0 else None
         )
         text_completion_only_loss = _text_completion_only_loss_arg(args)
+        text_assistant_only_loss = _text_assistant_only_loss_arg(args)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/trainer.py:3267]: Agreed, genuine. The is_vlm branch computed assistant_only_loss but never forwarded it to create_vlm_batches / iterate_vlm_training_batches (neither accepts it) nor failed fast, so a VLM run with assistant_only_loss=True silently trained the normal objective. TRL raises for this exact config. Fixed in f558caa0 with a fail-fast ValueError pointing to assistant_only_loss=False or train_on_responses_only.

File: unsloth_zoo/mlx/utils.py [hunk 6]
```diff
@@ -2488,6 +2490,659 @@ def collect_mlx_texts(target, item, *, dataset_text_field="text", is_vlm=False):
...
+    dataset_text_field="text",
+    formatting_func=None,
+    append_eos=True,
+    completion_only_loss=None,
+    assistant_only_loss=False,
+):
+    """Materialize text rows that need explicit labels into token ids."""
+    formatted = []
+    for item in dataset:
+        source = formatting_func(item) if formatting_func is not None else item
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py:3175]: Agreed. For a raw prompt/completion dataset with a formatting_func returning text and completion_only_loss left at the default, the formatter drops the prompt/completion boundary so the labeled branch yields no rows, and the old guard only raised for explicit completion_only_loss=True, so it fell through to full-sequence training. Fixed in f558caa0: added _prompt_completion_formatter_conflict wired into create_batches, create_ordered_batches and iterate_training_batches after the labeled branch produces no rows, so a formatter that preserves prompt/completion keys still yields labeled rows and is unaffected.

File: unsloth_zoo/mlx/utils.py [hunk 7]
```diff
@@ -3569,6 +4225,62 @@ def create_batches(dataset, tokenizer, batch_size, max_seq_length,
...
+            max_seq_length,
+            num_batches=num_batches,
+            seed=seed,
+            pad_id=_mlx_text_pad_id(tokenizer),
+        )
+    if assistant_only_loss:
+        _validate_mlx_text_assistant_only_dataset(dataset)
+
+    if completion_only_loss is not False or assistant_only_loss:
+        normalize_mlx_chat_template(
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py:4659]: TRL does not auto-swap in a generation-marker training template here. transformers apply_chat_template(return_assistant_tokens_mask=True) does not inject the markers, and TRL raises the same no-assistant-tokens error when the model template lacks them. The MLX raise is exact parity, so there is no automatic fallback to mirror. Leaving as-is.

**Comment by danielhanchen** [comment]
@codex review
